### DISCUSSION TITLE
OpenGL rendering on entire screen

### DISCUSSION THREAD

**Discussion by nulliousp** [ocornut/imgui#4590] [New Users Build/Link/Run issues ONLY!]
I have some code that is supposed to only make the window screen change colour, but the entire thing is, it also hides the top bar.

``` c++
static void draw_callback(const ImDrawList* parent_list, const ImDrawCmd* cmd)
{
	glViewport(0, 0, 300, 300);
	glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	glViewport(Core::GetScreenSize(window).x/2, Core::GetScreenSize(window).y/2, Core::GetScreenSize(window).x, Core::GetScreenSize(window).y);
}

int main() {
   while (!glfwWindowShouldClose(window)) {
        ImGui_ImplOpenGL3_NewFrame();
	ImGui_ImplGlfw_NewFrame();
	ImGui::NewFrame();

	// Dockspace
	ImGui::DockSpaceOverViewport(ImGui::GetMainViewport());
        ImGui::Begin("Viewport");

        ImGui::GetWindowDrawList()->AddCallback(draw_callback, NULL);

        ImGui::End();

        ImGui::Render();
	int w, h;
	glfwGetFramebufferSize(window, &w, &h);
	glViewport(0, 0, w, h);
	glClearColor(0.45f, 0.55f, 0.60f, 1.00f);
	glClear(GL_COLOR_BUFFER_BIT);
	ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());

	if (io.ConfigFlags & ImGuiConfigFlags_ViewportsEnable)
	{
		GLFWwindow* backup_current_context = glfwGetCurrentContext();
		ImGui::UpdatePlatformWindows();
		ImGui::RenderPlatformWindowsDefault();
		glfwMakeContextCurrent(backup_current_context);
	}

	/*End of rendering*/

	glfwPollEvents();
	glfwSwapBuffers(window);
    }
```

**Comment by PathogenDavid** [comment]
`glClear` is not affected by the current viewport.

If I remember right you can use `glScissor` instead (make sure you enable it with `glEnable(GL_SCISSOR_TEST)`.)

Edit: Also, why not just use `ImGuiDrawList::AddRectFilled`?

**Reply by nulliousp** [reply]
I am just testing it out. Just to see what I can do with it

**Reply by nulliousp** [reply]
Thanks! glScissor() worked!

**Reply by nulliousp** [reply]
Whenever I use ImGui::GetWindowPos() it shows in a different location. Can this be a problem with ImGui using the center as 0 instead of the bottom left?

**Reply by nulliousp** [reply]
The only reason I don't use ImGui::Image() as it never seems to work for me.

**Reply by PathogenDavid** [reply]
> Whenever I use ImGui::GetWindowPos() it shows in a different location.

I mean the coordinates are going to change as the window moves. I'm not sure if I understand what you're confused about.

> Can this be a problem with ImGui using the center as 0 instead of the bottom left?

`(0, 0)` is at the top-left of your window for single viewport applications.

For multi-viewport applications (IE: on the docking branch), `(0, 0)` is the top-left of the primary monitor. Use `ImGui::GetMainViewport()` or `ImGui::GetWindowviewport()` to get the relevant viewport and use the `Pos` to get the top-left of the primary viewport.

**Reply by nulliousp** [reply]
I don't know if this is right:

```c++
glScissor(ImGui::GetMainViewport()->Pos.x + vec2p.x, ImGui::GetMainViewport()->Pos.y + vec2p.y/2, vec2s.x, vec2s.y);
```

vec2p is the vector2 position of the imgui window nad the vec2s is the vector2 size of the imgui window.

It's not showing up in the right area no-matter whether it is *, /, + or -. I am confused

**Reply by PathogenDavid** [reply]
(Edit for context: If it feels like I'm responding to something else it's because AnnoyingB changed their entire comment.)

Are you trying to use `GetWindowPos` during your callback? If so you can't do that. `ImGui::GetWindowPos` gets the position of the current window. The callback isn't actually executed until later during rendering, at which time there isn't really* a current window.

(*I think technically the default debug window will be the current window, but I would not rely on that.)

You can either use `ImDrawCmd::ClipRect` to get the clipping rectangle of the window (note the coordinates will be flipped vertically in OpenGL), pass the window position indirectly through `callback_data`, or pass the (internal) value of `ImGui::GetCurrentWindow` through `callback_data`.

----------------

As an aside, I don't think what you're doing is a very great idea if you're planning to use this to render something complicated. You'd be better off figuring out why you can't get the `ImGui::Image` with a render target approach to work.

**Reply by nulliousp** [reply]
(Edit: I also spent a lot of time researching and trying things out to get it to work and nothing. Also, do you mean comment or reply?)

Well, I use GetWindowPos() before I actually use the callback (forgot to mention that) and I do not want to use ImGui::Image() as I have tried to get it to work for ages (possibly a few days) and I got the same result, a black image. I think ImGui::Image is a pain to get it to work so I just gave up on it.

**Reply by PathogenDavid** [reply]
Please don't replace your entire comment like that, it breaks the coherency of the conversation.

You need to be careful about what coordinate space you're working in. What I was talking about with the viewport stuff was describing Dear ImGui's coordinate space. When you're using `glScissor` you're dealing with OpenGL's window coordinate space. `0, 0` is the left-bottom corner in OpenGL window coordinate space.

**Reply by nulliousp** [reply]
Oh, I understand what is wrong. I was getting the window position, not the ImGui window position. Thanks for the help.

**Reply by PathogenDavid** [reply]
No problem, glad you figured it out!

It feels to me like you're missing some fundamentals of graphics programming in general. Being able to render to a render target texture (RTT) is a pretty fundamental use of a graphics API. I feel like if you can't do that, you're just delaying the inevitable for when you run into a different problem later.

Remove Dear ImGui from the picture: You should focus on making a small app to test your understanding of RTTs in OpenGL before you try to do everything all together with Dear ImGui. It's just giving you extra stuff to worry about.

* Make an app that renders a triangle to the screen. (Just your everyday average OpenGL "Hello Triangle" demo.)
* Make a different app that renders a spinning cube with some texture.
* Take the triangle rendering code from the first one and put it in the second one, but have the triangle scene render to an RTT instead of the screen.
* Change the cube to use that RTT as its texture.

If you can do all that you should be able to do the same in the context of Dear ImGui with relative ease ✨

**Reply by nulliousp** [reply]
Is an RRT basically an FBO in OpenGL? Or am I wrong?

**Reply by PathogenDavid** [reply]
Oops, yeah. Sorry for the confusion. Render target texture is what DirectX and most game engines call them, they're the same thing.

**Reply by nulliousp** [reply]
Okay, not sure if this is right but:

```c++
	float vertices[] = {
		0.0f, 0.0f,
		0.0f, 0.5f,
		1.0f, 0.0f
	};

	GLuint FBO;
	glGenBuffers(1, &FBO);
	glBindBuffer(GL_FRAMEBUFFER, FBO);
	glBufferData(GL_FRAMEBUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);

	glBindBuffer(GL_FRAMEBUFFER, 0);

        ImGui::Image((void*)(ImTextureID)FBO, ImGui::GetWindowSize(), ImVec2(0, 1), ImVec2(1, 0));
```

Not sure what is wrong.

**Reply by PathogenDavid** [reply]
Again, focus on getting things working in isolation before trying to combine everything together. Making tiny apps that do just one thing is a good way to learn how things work.

**Reply by nulliousp** [reply]
Okay. Thanks for the advice!

**Reply by nulliousp** [reply]
```
Take the triangle rendering code from the first one and put it in the second one, but have the triangle scene render to an RTT instead of the screen.
```

This is confusing and hard for me :\. I can't get anything to render onto an FBO. Could you help?

**Reply by PathogenDavid** [reply]
Can you share what you have so far as a [GitHub Gist](https://gist.github.com/)?

**Reply by nulliousp** [reply]
https://gist.github.com/AnnoyingB/545a70a872fe4dfaddf8a3be7da93a0d

I am using GLEW and GLFW.

**Reply by PathogenDavid** [reply]
Thanks! Although I was hoping you'd share your attempt at getting it to render to an FBO. Or are you not sure where to start in general?

**Reply by PathogenDavid** [reply]
Actually I think I can infer based on your earlier comment https://github.com/ocornut/imgui/discussions/4590#discussioncomment-1392576 (I must admit I didn't look at it very closely the other day.)

I don't think you can use `glGenBuffers`/`glBindBuffer` with FBOs like that. You need to use `glGenFrameBuffers`/`glBindFramebuffer` instead. It's been too long since I've learned OpenGL to recommend a specific tutorial, but at a glance this one looks pretty reasonable: https://learnopengl.com/Advanced-OpenGL/Framebuffers

(You might consider enabling OpenGL's debug context to catch these sorts of mistakes. See [the second half of this guide](https://learnopengl.com/In-Practice/Debugging) or [the OpenGL wiki](https://www.khronos.org/opengl/wiki/Debug_Output))

**Reply by nulliousp** [reply]
Ah, I have understood framebuffers. This is what I used: http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-14-render-to-texture/. Thanks for using your time to help me!

**Reply by PathogenDavid** [reply]
No problem, glad I could help!
